### Imports

In [73]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler

### Load Data

In [3]:
df = pd.DataFrame.from_csv("AArete_data.csv")

In [4]:
df

Age  Weight  Sex Eye_Color  Spend_Dollars
PERSON_ID                                           
1           22     168    M      Blue            224
2           24     220    M     Green            600
3           18     175    F       NaN           1000
4           65     140    M     Brown            116
5           44     134    F     Brown              0
6           56     187    F      Blue          16670
7           21     199    F      Blue            589
8           16     165    M       NaN             44
9           78     166    M      Blue           1155
10          87     500  NaN      Blue           5841

### Questions

Question 1: 
Create a binary variable for “Spend_Dollars” to a binary classifier that is &gt;=1000 and &lt;1000

In [7]:
df["above_below_1000"] = df.Spend_Dollars >= 1000

In [8]:
df

Age  Weight  Sex Eye_Color  Spend_Dollars above_below_1000
PERSON_ID                                                            
1           22     168    M      Blue            224            False
2           24     220    M     Green            600            False
3           18     175    F       NaN           1000             True
4           65     140    M     Brown            116            False
5           44     134    F     Brown              0            False
6           56     187    F      Blue          16670             True
7           21     199    F      Blue            589            False
8           16     165    M       NaN             44            False
9           78     166    M      Blue           1155             True
10          87     500  NaN      Blue           5841             True

Question 2: Write a pairwise correlation matrix for all numeric variables

In [10]:
df.corr()

Age    Weight  Spend_Dollars  above_below_1000
Age               1.000000  0.480046       0.363910          0.535086
Weight            0.480046  1.000000       0.264104          0.416622
Spend_Dollars     0.363910  0.264104       1.000000          0.582876
above_below_1000  0.535086  0.416622       0.582876          1.000000

Question 3:

From your binary variable from # 1, write a simple logistic regression model with the variables

Eye Color, Weight, Sex

a. How would you calculate an ROC Curve? Syntax not required. Keep answer 4-5

sentences

Create Dummy Variables

In [27]:
df_eye_color = pd.get_dummies(df["Eye_Color"])

In [29]:
df_sex = pd.get_dummies(df["Sex"])

In [30]:
df = pd.concat([df, df_eye_color, df_sex], axis=1, join='inner')

In [33]:
#new df
df

Age  Weight  Sex Eye_Color  Spend_Dollars above_below_1000  Blue  \
PERSON_ID                                                                     
1           22     168    M      Blue            224            False     1   
2           24     220    M     Green            600            False     0   
3           18     175    F       NaN           1000             True     0   
4           65     140    M     Brown            116            False     0   
5           44     134    F     Brown              0            False     0   
6           56     187    F      Blue          16670             True     1   
7           21     199    F      Blue            589            False     1   
8           16     165    M       NaN             44            False     0   
9           78     166    M      Blue           1155             True     1   
10          87     500  NaN      Blue           5841             True     1   

           Brown  Green  F  M  
PERSON_ID                      
1              0      0  0  1  
2              0      1  0  1  
3              0      0  1  0  
4              1      0  0  1  
5              1      0  1  0  
6              0      0  1  0  
7              0      0  1  0  
8              0      0  0  1  
9              0      0  0  1  
10             0      0  0  0

In [47]:
X = df[["Blue","Brown","M","Weight"]]
y = df["above_below_1000"]

In [48]:
#train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [49]:
# Create logistic regression object
logistic = linear_model.LogisticRegression()

# Train the model using the training sets
logistic.fit(X, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [50]:
# The coefficients
print('Coefficients: \n', logistic.coef_)

Coefficients: 
 [[ 0.29188707 -0.55024557 -0.68830048  0.00251134]]


ROC: 

A ROC plots the true positives versus the false positives and can help us set a threshold to maximize true positves while while minimizing false positives. To do this you determine the Sensitivity = TP/(TP + FN) and the Specificity = TN/(TN + FP). You then graph 1-Specificity on the x-axis and the Sensitivity on the y-axis. In a perfect scenario the TP will be 1 and the FP will be 0, this represented by the uper left corner of the graph. 

Question 4: Create a simple linear regression with your dependent variable as spend_dollars, and the

following variables, sex, weight. Assume your target is normally distributed

a. How would you interpret the coefficient for sex? For weight? Syntax not required

b. Suppose your target variable is right-skewed, what steps would you take to adjust your

model? Syntax not required, Keep answer to 4-5 sentences.

In [69]:
X_reg = df[["M","F","Weight"]]
y_reg = df["Spend_Dollars"]


In [70]:
# Create logistic regression object
reg = linear_model.LinearRegression()

# Train the model using the training sets
reg.fit(X_reg, y_reg)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [71]:
# The coefficients
print('Coefficients: \n', reg.coef_)

Coefficients: 
 [  8923.07679165  12974.84781619     43.68152587]


Interpreting coefficients:

Weight: This is a continuous variable so for every one unit increase in weight, we can expect the dollars spent to go up about $43.7

Sex: This is a categorical variable. So if the person is female, we can expect the dollars spent to go up 8923 and if male the amount spent will is expected to be 12974. (Typically, one dummy variable should be dropped to reduce the redundancy. If a person is male, you know they are not female, so we don't need the femal column)

Right Skew: if the model has a skew the data can be standardized to set the mean to 0 and with 1 unit variance. 

Question 5: Write a function to standardize your Spend_Dollars to a new variable

Spend_Dollars_Standardized

In [78]:
Spend_Dollars_Standardized = StandardScaler().fit_transform(df.Spend_Dollars)

/Users/anaelisagentle/anaconda/lib/python3.6/site-packages/sklearn/utils/validation.py:429: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, _DataConversionWarning)
/Users/anaelisagentle/anaconda/lib/python3.6/site-packages/sklearn/preprocessing/data.py:586: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  warnings.warn(DEPRECATION_MSG_1D, DeprecationWarning)
/Users/anaelisagentle/anaconda/lib/python3.6/site-packages/sklearn/preprocessing/data.py:649: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  warnings.warn(DEPRECATION_MSG_1D, DeprecationWar

In [79]:
Spend_Dollars_Standardized

array([-0.48360693, -0.40783869, -0.32723418, -0.50537015, -0.52874546,
        2.83044767, -0.41005532, -0.51987897, -0.29599993,  0.64828196])

6) Create a for loop to print each age

In [89]:
df.Age.apply(lambda x: x)

PERSON_ID
1     22
2     24
3     18
4     65
5     44
6     56
7     21
8     16
9     78
10    87
Name: Age, dtype: int64